### PR TITLE
Improve Odometry Output with New RPM output on Rover Pro; Maintain backwards compatibility with older FW Versions

### PR DESCRIPTION
**Pull request by jbrover**
Improve Odometry Output with New RPM output on Rover Pro; Maintain backwards compatibility with older FW Versions.

### CODE REVIEW

File: rr_openrover_driver/src/odom_control.cpp [hunk 1]
```diff
@@ -272,30 +276,38 @@ int OdomControl::deadbandOffset(int motor_speed, int deadband_offset)
...
   velocity_history_.insert(velocity_history_.begin(), velocity);
   velocity_history_.pop_back();
-
-  if (accel > MAX_ACCEL_CUTOFF_)
-  {
-    change_in_velocity = 0.5*dt*MAX_ACCEL_CUTOFF_;
-    velocity = velocity_filtered_history_[0] + change_in_velocity;
+  
+  if(firmwareBuildNumber == BUILD_NUMBER_WITH_GOOD_RPM_DATA){
+    std::cout << "switching correctly" << std::endl;
```
Comment (rotu) [inline rr_openrover_driver/src/odom_control.cpp]: did you mean to check this in? it looks like debug output

Comment (jbrover) [reply rr_openrover_driver/src/odom_control.cpp]: I'll cleanup the std::out prints. There are a few others within the PR as well.

File: rr_openrover_driver/src/openrover_driver.cpp [hunk 2]
```diff
@@ -807,45 +807,73 @@ void OpenRover::serialManager()
...
-    right_vel_measured_ = 0;
-  }
-  else if (motor_speeds_commanded_[RIGHT_MOTOR_INDEX_] > MOTOR_NEUTRAL)  // this sets direction of measured
-  {
-    right_vel_measured_ = odom_encoder_coef_ / right_enc;
-  }
-  else
-  {
-    right_vel_measured_ = -odom_encoder_coef_ / right_enc;
+    float lconversion = left_rpm;
```
Comment (rotu) [inline rr_openrover_driver/src/openrover_driver.cpp]: lconversion/rconversion seem like vestigial variables and have confusing names.

Comment (jbrover) [reply rr_openrover_driver/src/openrover_driver.cpp]: Needed a way to get the signed int rpm data into a float data type so the division would actually run. I suppose a cast would be better and to eliminate those variables altogether.

**Comment by rotu** [comment]
Looks alright to me but I haven't seen the corresponding firmware. In particular, is RPM *actually* an RPM? Or just some quantity proportional to the motor speed?

File: rr_openrover_driver/src/openrover_driver.cpp [hunk 3]
```diff
@@ -807,45 +807,73 @@ void OpenRover::serialManager()
...
+    
+    float rconversion = right_rpm;
+
+    std::cout << "l period:  " << lperiod << std::endl;
+    std::cout << "r period   " << rperiod << std::endl;
+
+    std::cout << "L rpm: " << left_rpm << std::endl;
+    std::cout << "R rpm: " << right_rpm << std::endl;
+
+    left_vel_measured_ = lconversion / 9100;
```
Comment (rotu) [inline rr_openrover_driver/src/openrover_driver.cpp]: 9100 is an unexplained magic number. Name it?

Comment (jbrover) [reply rr_openrover_driver/src/openrover_driver.cpp]: Done, added to constants.hpp

**Comment by jbrover** [comment]
> Looks alright to me but I haven't seen the corresponding firmware. In particular, is RPM _actually_ an RPM? Or just some quantity proportional to the motor speed?

RPM is very close to the actual RPM of the motor. This output of RPMs was verified against wheel rotations * gear ratio. It's not perfect but it's a good semi-instantaneous reading.

**Review by drhieu — CHANGES_REQUESTED** [review]
@shoemakerlevy9 
rr_openrover_driver/launch/starterkit_bringup.launch
Don't you waant those settings to be unchanged? and "closed_loop_control_on" to be false for production?

**Comment by jbrover** [comment]
> @shoemakerlevy9
> rr_openrover_driver/launch/starterkit_bringup.launch
> Don't you waant those settings to be unchanged? and "closed_loop_control_on" to be false for production?

We want customers to use the new rover pro firmware in combination with the new firmware.

**Comment by shoemakerlevy9** [comment]
- Build Number makes it so that we cant do a 7.0.0 release which undesirable but fine
- Lets changes the default for closed_loop_control_on back to FALSE to prevent possibility of runaway robot and let me sleep better at night.
- Lets test out the unboxing experience with 2WD kit and defaulting to 4WD (assuming user doesn't change from default before driving the robot)

**Review by shoemakerlevy9 — CHANGES_REQUESTED** [review]
Change open_loop_on back to FALSE in launch file

**Comment by jbrover** [comment]
Made changes to the launch file per the requests. This driver has been verified with firmware 1.0.7 for 4wd, 2wd, and flippers drive types.

**Pull request merged**